doc/design/proposal-synchronous-hooks.md

```diff
@@ -0,0 +1,331 @@
+
+```js
+function resolve(specifier, context, nextResolve) {
+  const resolved = nextResolve(specifier, context);
+  if (resolved.url.startsWith('http')) {
+    // This adds a fetch request to a pool
+    const syncResponse = startFetch(resolved.url);
+    responseMap.set(url, syncResponse);
+  }
+  return resolved;
+}
+
+function load(url, context, nextLoad) {
+  const syncResponse = responseMap.get(url);
+  if (syncResponse) {
+    const source = syncResponse.wait();
+    return { source, shortCircuit: true };
+  }
+  return nextLoad(url, context);
+}
+```
+
+However, to reuse `resolve` for concurrent loading, we need to implement it to be run in BFS order - that is, for a module like this:
+
+```js
+import 'a';
+import 'b';
+import 'c';
+```
+
+Instead of running the hooks as:
+
+1. `resolve(a)` -> `load(a)` -> `link(a)`
+2. `resolve(b)` -> `load(b)` -> `link(a)`
+3. `resolve(c)` -> `load(c)` -> `link(a)`
+
+We need to run the hook as:
+
+1. `resolve(a)` -> `resolve(b)` -> `resolve(c)` (starts the off-thread fetching requests concurrently)
+2. `load(a)` -> `load(b)` -> `load(c)` (block on the concurrently fetched results)
+3. `link(a)` -> `link(b)` -> `link(c)`
+
+Otherwise, we need to invent another hook that gets run in BFS order before both `resolve` and `load`, but in that case, the hook could only get the specifiers in its arguments, which may or may not be a problem. For network imports it's okay, because the specifiers are supposed to be URLs already, but other forms of concurrent processing (e.g. parallel transpilation with a worker pool) may not be possible without the result of `resolve`.
+
+On the other hand, for CJS we have to run the hooks in DFS order since that's how CJS module requests have to be resolved.
+
+## Child worker/process hook inheritance
+
+Unlike the old `module.register()` design, this new API takes an object with functions so that it's possible to migrate existing monkey-patching packages over. One reason for the `path`/`url` parameter in the `module.register()` and the `initialize` was to aid hook inheritance in child worker and processes. With the new API, we can compose it differently by adding a new API specifically for registering a script that should be preloaded by child workers or processes. Hook authors can either put the call together with the hook registration (so that the inheritance is recursive automatically for grandchildren etc.), or put the hook registration code in a different file and configure a preload of that file.
+
+```js
+process.preloadInChildren(__filename);  // Or `import.meta.filename`
```

Comment:
Is `preloadInChildren` recursive or does it only apply to the direct child threads?

If it is not recursive, then this risks a situation where worker threads created by worker threads might not work properly if hooks authors forget to have **both** the hooks registration and the `preloadInChildren` call in the same file.

For example someone might write:

index.js
```js
process.preloadInChildren(fileURLToPath(import.meta.resolve('./hooks.mjs'));
```

hooks.mjs
```js
module.registerHooks({ load });
```

and see that it works fine for workers and consider their job done. It is unlikely loaders authors will properly test worker threads of worker threads!

Having the registration be recursive seems more well-defined, and more similar to the for example the new browser spec for realms attenuation (https://github.com/WICG/Realms-Initialization-Control?tab=readme-ov-file#proposal), where this realms init function applies recursively to every context.

This makes for a much better security primitive, but would be incompatible with the code example here by being recursive, which would result in duplicate registrations. Alternatively is registration deduping being considered for this API?

I would really value a little more design clarity on these questions.

Reply:
If we really need to figure out worker inheritance, I think we should just exclude it in the MVP and don't discuss it in this proposal - I suspect trying to be clever about the inheritance would break existing CJS monkey-patchers, while this proposal is prioritized towards deprecating monkey-patching, not a full replacement of `module.register()`. The off-thread move in `module.register()` was already an attempt that tried to be clever and ended up with many breakages.

Reply:
Also, the worker inheritance topic is something that stretches beyond module hooks IMO. I think the risk of only figuring it out for module hooks but not the other use cases has a bigger risk in making the core APIs messier than necessary, compared to not figuring it out for module hooks for now and come up with something that works for all use cases later. If I have to choose I'd prefer making the core APIs less messy by not rushing a design that's based purely on module hooks use cases.

Reply:
The tradeoff goes both ways here though - if we don't think of the module harmony effort while building this design we risk a rewrite down the line again. As someone involved in that effort, this feedback very much comes out of the work we're doing on https://github.com/tc39/proposal-esm-phase-imports.

Reply:
If anything that seems like a more convincing argument to not include discussion about worker inheritance in the hooks proposal, as trying to be clever with some default hooks inheritance behavior is more likely to clash with any inheritance behavior that browsers come up with/decide later, compared to not doing anything special for now. It's less risky to just provide an API that's closer to what the ecosystem has been doing for many many years (e.g. pirates & require-in-the-middle). Whatever the spec does should also consider and accommodate ecosystem practices like what we try to do here IMO.

Reply:
Discussed in the meeting this week and we agreed that we should do function-based registration and leave automatic inheritance to a dedicated preload API. I summarized the reasons in https://github.com/nodejs/loaders/issues/203#issuecomment-2288306105 - will update the doc a bit later.